### PR TITLE
For Issue #1375 - Eval5 CARLA Object Detection AdversarialPatchPyTorch

### PR DESCRIPTION
**Pull request by yusong-tan**
Adding AdversarialPatchPyTorch with differentiable perspective transform and modifications to attack RGB (for single modality) and RGB+Depth (for multimodality)

### CODE REVIEW

**Comment by lcadalzo** [comment]
Confirmed that on a single RTX 6000 GPU I'm seeing ~7.5 iters/s compared to ~3.3 using RobustDPatch. Before I do so myself, @yusong-tan have you done any benchmarking to confirm that adversarial results are similar between the two attacks?

**Comment by yusong-tan** [comment]
@lcadalzo No, I have not done any metrics comparison other than speed. AdversarialPathPyTorch has differentiable perspective transform, and RobustDPatch doesn't so the number of attack iterations needed for comparison will be different.

**Comment by lcadalzo** [comment]
Interestingly I'm seeing far lower adversarial mAP using `AdversarialPatchPyTorch` compared to `RobustDPatch` when testing the undefended RGB-only baseline model on the `"large"` split of the test set. As noted in the scenario docs, the RobustDPatch adversarial mAP and hallucinations/img for the undefended model are 0.17 and 10.3 respectively.

When I run the config in this PR, `carla_obj_det_adversarialpatch_undefended.json`, I see adversarial mAP and hallucinations/img of 0.05 and 31.6 respectively. My intuition would be that this is due to the fact that perspective transform is now differentiable

**Comment by yusong-tan** [comment]
@lcadalzo Your observation is similar to mine. I suspect to get comparable results, Robust DPatch may need many more attack iterations.

When we have a PyTorch defended model, we may want to also attack it once with Robust DPatch to provide an idea of the degradation from using it.

**Comment by lcadalzo** [comment]
There's a lot of attack code here and much of it seems to be copied from ART. If possible, I'm looking to see if there are ways to reduce the amount of copied code. Do you have thoughts in this direction?

**Comment by lcadalzo** [comment]
One major example is the `_get_losses()` method that's added. It appears to be identical to ART except for [this](https://github.com/twosixlabs/armory/blob/5a3e67f486f8e6d36ee793d2119bd196d91c6562/armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py#L103) line which allows a batch dim of 6. I wonder if this is something that could be modified in ART instead

**Comment by lcadalzo** [comment]
And if we need to modify `_get_losses()`, I think it might be cleaner to do so where the baseline model is defined (as opposed to in the attack code) since this is a model-specific change. Patching the method in the attack code means that anyone who uses this attack will have their model updated with this method

**Comment by yusong-tan** [comment]
@lcadalzo I prefer to make the one line `_get_losses()` change to be in ART, but I don't think we can get that change into the upcoming Armory release. If you think this is possible, we can reach out to Beat.

I agree that `_get_losses()` should be in the baseline model. It was an error on my part.

**Comment by lcadalzo** [comment]
For my understanding: what's the motivation behind averaging the gradient over the depth channels [here](https://github.com/yusong-tan/armory/blob/eval5_carla_obj_det_adversarialpatchpytorch/armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py#L191-L193)?

Additionally, why are the depth-channel gradients divided by `self.learning_rate * self.learning_rate_depth` [here](https://github.com/yusong-tan/armory/blob/eval5_carla_obj_det_adversarialpatchpytorch/armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py#L198)?

**Comment by lcadalzo** [comment]
> @lcadalzo I prefer to make the one line `_get_losses()` change to be in ART, but I don't think we can get that change into the upcoming Armory release. If you think this is possible, we can reach out to Beat.

I've reached out to Beat to see whether this is possible

File: armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+                    max=self.estimator.clip_values[1],
+                )
+
+                if self._patch.shape[0] == 6:
+                    self._patch[3:, :, :] = torch.clamp(
+                        self._patch[3:, :, :], min=self.min_depth, max=self.max_depth
+                    )
+        else:
+            raise ValueError(
+                "Adam optimizer for CARLA Adversarial Patch not supported."
```
Comment (lcadalzo) [inline armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py:214]: is there meant to be modified code below the `ValueError` that won't be reached? If the only possible values for `self._optimizer_string` are `"pgd"` and `"Adam"`, then it seems like the following would make the most sense:

```
if "pgd":
    ...
elif "Adam":
    raise the ValueError not supported for CARLA
else:
    raise a ValueError for unrecognized _optimizer_string
```

File: armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+        import torch  # lgtm [py/repeated-import]
+
+        self.estimator.model.zero_grad()
+        loss = self._loss(images, target, mask)
+        loss.backward(retain_graph=True)
+
+        if self._optimizer_string == "pgd":
+            patch_grads = self._patch.grad
+            if patch_grads.shape[0] == 6:
+                patch_grads[3:, :, :] = torch.mean(
```
Comment (lcadalzo) [inline armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py:191]: what's the motivation for averaging over the depth-channel gradients?

File: armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+                    patch_grads[3:, :, :], dim=0, keepdim=True
+                )
+
+            gradients = patch_grads.sign() * self.learning_rate
+            if patch_grads.shape[0] == 6:
+                gradients[3:, :, :] = (
+                    gradients[3:, :, :] / self.learning_rate * self.learning_rate_depth
```
Comment (lcadalzo) [inline armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py:198]: why are depth-channel gradients divided by `self.learning_rate * self.learning_rate_depth`?

File: armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py (hunk 2 above)
Comment (yusong-tan) [reply armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py:191]: Depth only has one channel, but we repeat three times for it to work with the model on the forward pass. On the backward pass, the gradients are calculated for three depth channels, but we need to change them into three identical channels. To me, it makes more sense to average the gradients in the three depth channels than to randomly pick gradients for one of the channels.

File: armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py (hunk 1 above)
Comment (yusong-tan) [reply armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py:214]: Initially, when I used `self.learning_rate` for updating both RGB and depth, the Adam optimizer can be used. Later when I decided that the depth channel needed a separate learning rate, i.e., `self.learning_rate_depth`, the Adam optimizer could no longer accommodate two different learning rates. That's why I added the `raise ValueError()` but forgot to delete everything after it.

**Comment by lcadalzo** [comment]
I've created the following issue in ART to support 6-channel inputs: https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/1630. Per Beat, this can go in ART 1.10.1, and I'm waiting to hear when that will be released

**Comment by lcadalzo** [comment]
For the multimodal attack, no surprise but I'm seeing a much more successful attack on the "large" split of the test set: adversarial mAP of 0.08 and 42.6 hallucinations per image. The RobustDPatch RGB-only attack yielded an adversarial mAP of 0.42 and 1.7 hallucinations per image
